### PR TITLE
New resources - `azurerm_key_vault_managed_storage_account` and `azurerm_key_vault_managed_storage_account_sasdefinition`

### LINKED ISSUE #4659: Support for KeyVault Managed Storage SAS Definition
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Azure Key Vault supports creation of a AzKeyVaultManagedStorageAccount & AzKeyVaultManagedStorageSasDefinition. It would be great if the AzureRM terraform provider supported this too.   This is needed to implement the same steps as outlined in https://docs.microsoft.com/en-us/azure/key-vault/key-vault-overview-storage-keys-powershell

### New or Affected Resource(s)

* azurerm_key_vault_managed_storage_account
* azurerm_key_vault_managed_storage_sasdefinition

### Potential Terraform Configuration

```hcl
resource "azurerm_resource_group" "testrg" {
  name     = "resourceGroupName"
  location = "westus"
}

resource "azurerm_storage_account" "testsa" {
  name                     = "storageaccountname"
  resource_group_name      = "${azurerm_resource_group.testrg.name}"
  location                 = "westus"
  account_tier             = "Standard"
  account_replication_type = "GRS"

  tags = {
    environment = "staging"
  }
}

data "azurerm_storage_account_sas" "test" {
  connection_string = "${azurerm_storage_account.testsa.primary_connection_string}"
  https_only        = true

  resource_types {
    service   = true
    container = false
    object    = false
  }

  services {
    blob  = true
    queue = false
    table = false
    file  = false
  }

  start  = timestamp()
  expiry = timeadd(timestamp(), "24h")

  permissions {
    read    = true
    write   = true
    delete  = false
    list    = false
    add     = true
    create  = true
    update  = false
    process = false
  }
}

resource "azurerm_key_vault" "test" {
  name                = "testvault"
  location            = "${azurerm_resource_group.test.location}"
  resource_group_name = "${azurerm_resource_group.test.name}"

  sku_name = "standard"

  tenant_id = "22222222-2222-2222-2222-222222222222"

  enabled_for_disk_encryption = true

  tags = {
    environment = "staging"
  }

}

resource "azurerm_key_vault_access_policy" "test" {
  key_vault_id = "${azurerm_key_vault.test.id}"

  tenant_id = "00000000-0000-0000-0000-000000000000"
  object_id = "ae8494f6-67f2-4e56-b015-041a9ca4d0e8"

  storage_permissions = [
    "get",
    "list",
    "delete",
    "set",
    "update",
    "regeneratekey",
    "getsas",
    "listsas",
    "deletesas",
    "setsas",
    "recover",
    "backup",
    "restore",
    "purge"
  ]

}

resource "azurerm_key_vault_managed_storage_account" "test" {
  key_vault_id = "${azurerm_key_vault.test.id}"
  storage_account_id  = "${azurerm_storage_account.testsa.id}"
  storage_account_key = "key1"
  auto_regenerate_key = false
  regeneration_period_days = 1
}

resource "azurerm_key_vault_managed_storage_sasdefinition" "test" {
  key_vault_id = "${azurerm_key_vault.test.id}"
  storage_account_id  = "${azurerm_storage_account.testsa.id}"
  sas_definition_name = "ExpireDaily"
  sas_template_uri = "${azurerm_storage_account_sas.test.sas}"

}
```

### References

https://docs.microsoft.com/en-us/azure/key-vault/key-vault-overview-storage-keys-powershell#shared-access-signature-tokens

### PR DESCRIPTION
**Pull request by MattiasAng**
Fixes #4659.

### CODE REVIEW

File: azurerm/internal/services/keyvault/key_vault_managed_storage_account.go [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+				log.Printf("[DEBUG] Managed Storage Account %q recovered with ID: %q", name, *recoveredStorageAccount.ID)
+
+				_, err := client.SetStorageAccount(ctx, *keyVaultBaseUrl, name, parameters)
+				if err != nil {
+					return err
+				}
+			}
+		} else {
+			// If the error response was anything else, or `recover_soft_deleted_key_vaults` is `false` just return the error
+			return err
```
Comment (katbyte) [inline azurerm/internal/services/keyvault/key_vault_managed_storage_account.go:169]: could we wrap this error?

File: azurerm/internal/services/keyvault/key_vault_managed_storage_account.go [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+
+				if _, err := stateConf.WaitForState(); err != nil {
+					return fmt.Errorf("Error waiting for Managed Storage Account %q to become available: %s", name, err)
+				}
+				log.Printf("[DEBUG] Managed Storage Account %q recovered with ID: %q", name, *recoveredStorageAccount.ID)
+
+				_, err := client.SetStorageAccount(ctx, *keyVaultBaseUrl, name, parameters)
+				if err != nil {
+					return err
+				}
```
Comment (katbyte) [inline azurerm/internal/services/keyvault/key_vault_managed_storage_account.go:165]: we can combine these lines? and wrap this error
```suggestion
				if _, err := client.SetStorageAccount(ctx, *keyVaultBaseUrl, name, parameters); err != nil {
					return err
				}
```

File: azurerm/internal/services/keyvault/key_vault_managed_storage_account_sasdefinition.go [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+				log.Printf("[DEBUG] Managed Storage Account Sas Definition %q recovered with ID: %q", name, *recoveredStorageAccount.ID)
+
+				_, err := client.SetSasDefinition(ctx, *keyVaultBaseUri, storageAccount.Name, name, parameters)
+				if err != nil {
+					return err
+				}
+			}
+		} else {
+			// If the error response was anything else, or `recover_soft_deleted_key_vaults` is `false` just return the error
+			return err
```
Comment (katbyte) [inline azurerm/internal/services/keyvault/key_vault_managed_storage_account_sasdefinition.go:171]: same comments as above

File: azurerm/internal/services/keyvault/registration.go [hunk 4]
```diff
@@ -35,12 +35,14 @@ func (r Registration) SupportedDataSources() map[string]*schema.Resource {
...
-		"azurerm_key_vault":                                  resourceKeyVault(),
+		"azurerm_key_vault_access_policy":                         resourceKeyVaultAccessPolicy(),
+		"azurerm_key_vault_certificate":                           resourceKeyVaultCertificate(),
+		"azurerm_key_vault_certificate_issuer":                    resourceKeyVaultCertificateIssuer(),
+		"azurerm_key_vault_key":                                   resourceKeyVaultKey(),
+		"azurerm_key_vault_managed_hardware_security_module":      resourceKeyVaultManagedHardwareSecurityModule(),
+		"azurerm_key_vault_secret":                                resourceKeyVaultSecret(),
+		"azurerm_key_vault":                                       resourceKeyVault(),
+		"azurerm_key_vault_managed_storage_account":               resourceKeyVaultManagedStorageAccount(),
+		"azurerm_key_vault_managed_storage_account_sasdefinition": resourceKeyVaultManagedStorageAccountSasToken(),
```
Comment (katbyte) [inline azurerm/internal/services/keyvault/registration.go:46]: this name seems incomplete, should we make it?
```suggestion
		"azurerm_key_vault_managed_storage_account_sas_token_definition": resourceKeyVaultManagedStorageAccountSasTokenDefinition(),
```

**Review by katbyte — CHANGES_REQUESTED** [review]
Thanks @MattiasAng - i've given this an initial review and left some comments inline to address

**Comment by katbyte** [comment]
Hey @MattiasAng - as theres been no activity on this for a couple months now i'm going to close this PR - thanks for opening it and if you do come back to address the comments please do re open it!

**Comment by tjololo** [comment]
@katbyte @MattiasAng seems like we are close to a solution for #4659.
Any chance we can make this happen?

**Comment by aristosvo** [comment]
@MattiasAng If you don't mind, I'll try to rebase and close this in another PR.

**Comment by MattiasAng** [comment]
Thanks @aristosvo, too much at work to finish this one.

**Pull request closed without merging**